File: backend/app/functions/fetch_assets_config.py
import requests, json, urllib.parse
from flask import request, redirect, session
from app.config import config
from pathlib import Path

def fetch_assets_config():
    access_token = request.args.get("access_token")
    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json",
        "Accept": "application/json"
    }

    project_id = config.project_id
    base_url = "https://developer.api.autodesk.com/construction/assets/v1/projects"

    try:
        # === 1️⃣ Fetch custom attributes ===
        ca_resp = requests.get(f"{base_url}/{project_id}/custom-attributes", headers=headers)
        custom_attributes = ca_resp.json().get("results", []) if ca_resp.status_code == 200 else []

        # === 2️⃣ Fetch status-step sets ===
        ss_resp = requests.get(f"{base_url}/{project_id}/status-step-sets", headers=headers)
        status_sets_raw = ss_resp.json().get("results", []) if ss_resp.status_code == 200 else []

        status_sets = []
        status_sets_dict = {}
        for ss in status_sets_raw:
            ss_obj = {
                "id": ss.get("id"),
                "name": ss.get("name"),
                "description": ss.get("description", ""),
                "statuses": ss.get("values", [])
            }
            status_sets.append(ss_obj)
            status_sets_dict[ss.get("id")] = ss_obj

        # === 3️⃣ Fetch raw categories ===
        cat_resp = requests.get(f"{base_url}/{project_id}/categories", headers=headers)
        categories_raw = cat_resp.json().get("results", []) if cat_resp.status_code == 200 else []

        # === 4️⃣ Fetch category → status set mappings using batch API ===
        category_ids = [cat["id"] for cat in categories_raw]
        batch_url = f"{base_url}/{project_id}/category-status-step-sets/status-step-sets:batch-get"
        batch_payload = {"ids": category_ids}
        print("Fetching category to status set mappings via batch API...", batch_payload)
        batch_resp = requests.post(batch_url, headers=headers, json=batch_payload)
        category_status_raw = batch_resp.json().get("results", []) if batch_resp.status_code == 200 else []

        # Build a mapping: categoryId -> statusStepSetId
        category_to_statusSet = {c["categoryId"]: c.get("statusStepSetId") for c in category_status_raw}

        # === 5️⃣ Map categories with their status sets and custom attributes ===
        categories_with_status = []
        for cat in categories_raw:
            status_set_id = category_to_statusSet.get(cat["id"])
            status_set = status_sets_dict.get(status_set_id)

            # Determine display name for root/system categories
            status_set_name = status_set.get("name") if status_set else "No Status Set (system)"
            statuses = status_set.get("statuses", []) if status_set else []

            category_custom_attributes = [
                {
                    "displayName": ca.get("displayName"),
                    "description": ca.get("description"),
                    "dataType": ca.get("dataType"),
                    "enumValues": ca.get("enumValues", []),
                }
                for ca in custom_attributes
            ]

            categories_with_status.append({
                "categoryName": cat.get("name"),
                "categoryId": cat.get("id"),
                "parentId": cat.get("parentId"),
                "statusSetId": status_set_id,
                "statusSetName": status_set_name,
                "statuses": statuses,
                "customAttributes": category_custom_attributes,
                "children": cat.get("subcategoryIds", [])
            })

        # === 6️⃣ Aggregate data ===
        aggregated_data = {
            "customAttributes": [{ca.get("displayName"): ca.get("description", "")} for ca in custom_attributes],
            "statusSets": status_sets,
            "categories": categories_with_status
        }

        # Store in session
        session["aggregated_data"] = aggregated_data

        print(aggregated_data)

        # === 7️⃣ Save all raw and mapped data to log.json ===
        log_path = Path("log.json")
        with log_path.open("w", encoding="utf-8") as f:
            json.dump({
                "customAttributesRaw": custom_attributes,
                "statusSetsRaw": status_sets_raw,
                "categoriesRaw": categories_raw,
                "categoryStatusMappingRaw": category_status_raw,
                "categoriesWithStatus": categories_with_status
            }, f, indent=2)

        print("\n=== Saved log.json with all debug info ===")
        print(log_path.resolve())

        # Redirect message
        msg = urllib.parse.quote_plus(
            f"Fetched {len(custom_attributes)} attributes, "
            f"{len(status_sets)} status sets, "
            f"{len(categories_raw)} categories"
        )
        return redirect(f"/?msg={msg}")

    except Exception as e:
        msg = urllib.parse.quote_plus(f"Error fetching assets info: {e}")
        return redirect(f"/?msg={msg}")

File: rootTS/src/functions/fetchAssetsConfig.ts
import axios from "axios";
import fs from "fs";
import path from "path";
import { Request, Response } from "express";
import { stringify } from "csv-stringify/sync";
import { config } from "../config";
import { getCsvPath, getOutputDir, writeCsv } from "../utils"; // Assuming writeCsv is exported, or I will use local helper below

// Helper to match Python's write_csv behavior (UTF-8 SIG for Excel)
function saveCsv(filename: string, data: any[], columns: string[]) {
  const csvPath = getCsvPath(filename);
  const output = stringify(data, {
    header: true,
    columns: columns,
    bom: true, // Matches encoding='utf-8-sig'
  });
  fs.writeFileSync(csvPath, output);
}

export async function fetchAssetsConfig(
  req: Request,
  res: Response,
  accessToken: string
) {
  const projectId = config.project_id;
  const baseUrl =
    "https://developer.api.autodesk.com/construction/assets/v1/projects";

  const headers = {
    Authorization: `Bearer ${accessToken}`,
    "Content-Type": "application/json",
    Accept: "application/json",
  };

  try {
    // === 1️⃣ Fetch custom attributes ===
    let customAttributes: any[] = [];
    try {
      const caResp = await axios.get(
        `${baseUrl}/${projectId}/custom-attributes`,
        { headers }
      );
      if (caResp.status === 200) customAttributes = caResp.data.results || [];
    } catch (e) {
      console.error("Error fetching custom attributes", e);
    }

    // === 2️⃣ Fetch status-step sets ===
    let statusSetsRaw: any[] = [];
    try {
      const ssResp = await axios.get(
        `${baseUrl}/${projectId}/status-step-sets`,
        { headers }
      );
      if (ssResp.status === 200) statusSetsRaw = ssResp.data.results || [];
    } catch (e) {
      console.error("Error fetching status sets", e);
    }

    const statusSets: any[] = [];
    const statusSetsDict: Record<string, any> = {};

    for (const ss of statusSetsRaw) {
      const ssObj = {
        projectId: ss.projectId,
        id: ss.id,
        name: ss.name,
        description: ss.description || "",
        statuses: ss.values || [],
      };
      statusSets.push(ssObj);
      statusSetsDict[ss.id] = ssObj;
    }

    // === 3️⃣ Fetch raw categories ===
    let categoriesRaw: any[] = [];
    try {
      const catResp = await axios.get(`${baseUrl}/${projectId}/categories`, {
        headers,
      });
      if (catResp.status === 200) categoriesRaw = catResp.data.results || [];
    } catch (e) {
      console.error("Error fetching categories", e);
    }

    // === 4️⃣ Fetch category → status set mappings using batch API ===
    const categoryIds = categoriesRaw.map((cat: any) => cat.id);
    const batchUrl = `${baseUrl}/${projectId}/category-status-step-sets/status-step-sets:batch-get`;
    const batchPayload = { ids: categoryIds };

    console.log(
      "Fetching category to status set mappings via batch API...",
      batchPayload
    );

    let categoryStatusRaw: any[] = [];
    try {
      const batchResp = await axios.post(batchUrl, batchPayload, {
        headers,
        params: { includeInherited: "true" },
      });
      if (batchResp.status === 200)
        categoryStatusRaw = batchResp.data.results || [];
    } catch (e) {
      console.error("Error batch fetching mappings", e);
    }

    console.log("GOTCHA");
    // Build a mapping: categoryId -> statusStepSetId
    const categoryToStatusSet: Record<string, string> = {};
    for (const c of categoryStatusRaw) {
      categoryToStatusSet[c.categoryId] = c.statusStepSetId;
    }

    // === 5️⃣ Map categories with their status sets and custom attributes ===
    const categoriesWithStatus: any[] = [];
    for (const cat of categoriesRaw) {
      const statusSetId = categoryToStatusSet[cat.id];
      const statusSet = statusSetsDict[statusSetId];

      // Determine display name for root/system categories
      const statusSetName = statusSet
        ? statusSet.name
        : "No Status Set (system)";
      const statuses = statusSet ? statusSet.statuses : [];

      const categoryCustomAttributes = customAttributes.map((ca: any) => ({
        displayName: ca.displayName,
        description: ca.description,
        dataType: ca.dataType,
        enumValues: ca.enumValues || [],
      }));

      categoriesWithStatus.push({
        categoryName: cat.name,
        categoryId: cat.id,
        parentId: cat.parentId,
        statusSetId: statusSetId,
        statusSetName: statusSetName,
        statuses: statuses,
        customAttributes: categoryCustomAttributes,
        children: cat.subcategoryIds || [],
      });
    }

    // === 6️⃣ Aggregate data ===
    const aggregatedData = {
      customAttributes: customAttributes.map((ca: any) => ({
        [ca.displayName]: ca.description || "",
        id: ca.id,
      })),
      statusSets: statusSets,
      categories: categoriesWithStatus,
    };

    // Store in session
    if (req.session) {
      (req.session as any).aggregated_data = aggregatedData;
    }

    // === 7️⃣ Save all raw and mapped data to log.json ===
    const logPath = path.join(getOutputDir(), "log.json");
    fs.writeFileSync(
      logPath,
      JSON.stringify(
        {
          customAttributesRaw: customAttributes,
          statusSetsRaw: statusSetsRaw,
          categoriesRaw: categoriesRaw,
          categoryStatusMappingRaw: categoryStatusRaw,
          categoriesWithStatus: categoriesWithStatus,
        },
        null,
        2
      )
    );

    console.log("\n=== Saved log.json with all debug info ===");
    console.log(logPath);

    // === 8️⃣ Save aggregated data to log_aggregated.json ===
    const logAggPath = path.join(getOutputDir(), "log_aggregated.json");
    fs.writeFileSync(logAggPath, JSON.stringify(aggregatedData, null, 2));
    console.log("\n=== Saved log_aggregated.json with aggregated data ===");

    const msg = encodeURIComponent(
      `Fetched ${customAttributes.length} attributes, ` +
        `${statusSets.length} status sets, ` +
        `${categoriesRaw.length} categories`
    );

    // === 8️⃣ Save useful data to CSVs ===

    // --- Status Sets CSV ---
    const statusSetsCsv: any[] = [];
    for (const ss of statusSets) {
      for (const status of ss.statuses) {
        statusSetsCsv.push({
          project_id: ss.projectId,
          status_set_id: ss.id,
          status_set_name: ss.name,
          status_id: status.id,
          status_label: status.label,
          status_description: status.description || "",
        });
      }
    }

    saveCsv("status_sets.csv", statusSetsCsv, [
      "project_id",
      "status_set_id",
      "status_set_name",
      "status_id",
      "status_label",
      "status_description",
    ]);
    console.log("\n=== Saved status_sets.csv with status sets data ===");

    // --- Custom Fields CSV ---
    const customFieldsCsv: any[] = [];
    for (const ca of customAttributes) {
      // keep semicolons inside the cell to separate multiple values
      let valuesRaw = "";
      if (ca.dataType === "enum") {
        const ev = ca.enumValues || [];
        // enumValues can be strings or objects; stringify safely
        valuesRaw = ev
          .map((v: any) => (typeof v === "string" ? v : String(v)))
          .join(";");
      }

      let valuesAndIdsRaw = "";
      if (ca.values) {
        valuesAndIdsRaw = (ca.values || [])
          .map((value: any) => {
            const disp = (value.displayName || "").trim();
            const id = (value.id || "").trim();
            return `${disp}(${id})`;
          })
          .join(";");
      }

      customFieldsCsv.push({
        project_id: ca.projectId,
        custom_attribute_id: ca.id,
        name: ca.name,
        display_name: ca.displayName,
        description: ca.description || "",
        data_type: ca.dataType,
        required: ca.requiredOnIngress ? "True" : "False",
        values: valuesRaw,
        values_and_ids: valuesAndIdsRaw,
      });
    }

    saveCsv("custom_fields.csv", customFieldsCsv, [
      "project_id",
      "custom_attribute_id",
      "name",
      "display_name",
      "description",
      "data_type",
      "required",
      "values",
      "values_and_ids",
    ]);
    console.log(
      "\n=== Saved custom_fields.csv with custom attributes data ==="
    );

    // --- Categories CSV ---
    const categoriesCsv: any[] = [];
    for (const cat of categoriesWithStatus) {
      const customAttrStr = (cat.customAttributes || [])
        .map((ca: any) => ca.displayName)
        .filter((n: any) => n)
        .join(";");

      categoriesCsv.push({
        project_id: projectId,
        category_id: cat.categoryId,
        category_name: cat.categoryName,
        parent_id: cat.parentId,
        status_set_id: cat.statusSetId,
        status_set_name: cat.statusSetName,
        custom_attributes: customAttrStr,
      });
    }

    saveCsv("categories.csv", categoriesCsv, [
      "project_id",
      "category_id",
      "category_name",
      "parent_id",
      "status_set_id",
      "status_set_name",
      "custom_attributes",
    ]);
    console.log("\n=== Saved categories.csv with categories data ===");

    // === 9️⃣ Create category_status_default.csv ===

    // Find IFCGlobalId custom attribute name
    let ifcGlobalIdName = "";
    for (const ca of customAttributes) {
      if ((ca.displayName || "").trim() === "IFCGlobalId") {
        ifcGlobalIdName = ca.name;
        break;
      }
    }

    // Build a mapping of status_set_name -> first status_id
    const statusSetNameToFirstStatus: Record<string, string> = {};
    for (const row of statusSetsCsv) {
      const ssName = row.status_set_name;
      const sId = row.status_id;

      // Keep the first status_id for each status_set_name
      if (ssName && sId && !statusSetNameToFirstStatus[ssName]) {
        statusSetNameToFirstStatus[ssName] = sId;
      }
    }

    // Build category_status_default.csv
    const categoryStatusDefaultCsv: any[] = [];
    for (const cat of categoriesCsv) {
      const ssName = cat.status_set_name;
      const defaultStatusId = statusSetNameToFirstStatus[ssName] || "";

      categoryStatusDefaultCsv.push({
        category_name: cat.category_name,
        category_id: cat.category_id,
        default_status_id: defaultStatusId,
        IFCGlobalID_cat_name: ifcGlobalIdName || "",
      });
    }

    // Write category_status_default.csv
    const outputDefPath = path.join(
      getOutputDir(),
      "category_status_default.csv"
    );
    // Using explicit BOM handling for Excel compatibility
    const outputDef = stringify(categoryStatusDefaultCsv, {
      header: true,
      columns: [
        "category_name",
        "category_id",
        "default_status_id",
        "IFCGlobalID_cat_name",
      ],
      bom: true,
    });
    fs.writeFileSync(outputDefPath, outputDef);

    console.log(
      `\n=== Saved category_status_default.csv with default status mappings to ${outputDefPath} ===`
    );
    console.log(decodeURIComponent(msg));

    return res.redirect(`/?msg=${msg}`);
  } catch (e: any) {
    const errMsg = encodeURIComponent(
      `Error fetching assets info: ${e.message}`
    );
    return res.redirect(`/?msg=${errMsg}`);
  }
}
